src/vdaf.rs

```diff
@@ -326,6 +343,7 @@ pub(crate) fn run_vdaf<V, M>(
 ) -> Result<V::AggregateResult, VdafError>
 where
     V: Client + Aggregator + Collector,
+    for<'a> &'a V::AggregateShare: Into<Vec<u8>>,
```

Comment:
Why is this bound required here? `run_vdaf` exercises agg share serialization. (Same comment below.)

Reply:
> I'm a big fan of the minimal API changes here. A couple high-level questions:
> 
> * It loos like the implementations of `From<&OutputShare<F>>` and `From<&AggregateShare<F>>` for `Vec<u8>` are identical. Does it make sense to refactor this so that they share the same implementation, then maybe `#[inline]` the implementation?
>   
>   * EDIT similar comment for `TryFrom<...>`.
> * Do any of our tests exercise this serialization logic? If not, I think we might as well add some basic functionality tests.

* I pulled the new `[Try]From` implementations to a common function; TBH, I was a little happier before (less indirection led to easier code reading IMO), but I definitely understand the desire to not repeat ourselves. Note I had to implement `From<Vec<F>>` for `OutputShare<F>` to make this work out.

* Not AFAICT, I added a couple of quick tests for round-tripping.

Reply:
> I'm a big fan of the minimal API changes here. A couple high-level questions:
> 
> * It loos like the implementations of `From<&OutputShare<F>>` and `From<&AggregateShare<F>>` for `Vec<u8>` are identical. Does it make sense to refactor this so that they share the same implementation, then maybe `#[inline]` the implementation?
>   
>   * EDIT similar comment for `TryFrom<...>`.
> * Do any of our tests exercise this serialization logic? If not, I think we might as well add some basic functionality tests.

* I pulled the new `[Try]From` implementations to a common function; TBH, I was a little happier before (less indirection led to easier code reading IMO), but I definitely understand the desire to not repeat ourselves. Note I had to implement `From<Vec<F>>` for `OutputShare<F>` to make this work out.

* Not AFAICT, I added a couple of quick tests for round-tripping.

Reply:
It's required because of the bound that was on Vdaf:
```
where
    for<'a> &'a Self::AggregateShare: Into<Vec<u8>>,
```

For reasons I don't fully understand, this constraint becomes "sticky" & must be re-stated everywhere a generic `Vdaf` parameter is used. See https://stackoverflow.com/questions/50090578/how-to-write-a-trait-bound-for-a-reference-to-an-associated-type-on-the-trait-it for more information.

Doing better will require a feature called "generic associated types" (per the SO above). Unfortunately, GATs are not stabilized yet; it looks like they are set to be stabilized in a few months (https://github.com/rust-lang/rust/issues/44265#issuecomment-1066173042).

Other than GATs, we can also state the required constraints in terms of a new trait, which allows us to refer to references to the associated types without the need for trait bounds on references to the associated types. While the new trait would be simple enough & would be easy to refactor into equivalent From/Into implementations once GATs are stable, I avoided doing so because this is more complex than copying the trait bound everywhere until GATs land.

FTR, the new trait would look something like:
```
/// ByteConvertible allows for a type to be encoded into bytes, and attempt to be decoded from
/// bytes.
pub trait ByteConvertible: Sized {
    /// The type of the error returned by try_from_bytes.
    type Error;

    /// Converts a value to an equivalent byte sequence.
    fn to_bytes(&self) -> Vec<u8>;

    /// Attempts to convert a byte sequence into a value.
    fn try_from_bytes<B: AsRef<[u8]>>(bytes: B) -> Result<Self, Self::Error>;
}
```